flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java

```diff
@@ -247,6 +256,12 @@
 	/** The total number of vertices currently in the execution graph. */
 	private int numVerticesTotal;
 
+	private final PartitionReleaseStrategy.Factory partitionReleaseStrategyFactory;
+
+	private PartitionReleaseStrategy partitionReleaseStrategy;
+
+	private SchedulingTopology schedulingTopology;
```

Comment:
How about initializing `schedulingTopology` and `partitionReleaseStrategy` in the constructor, then we could make them final, and no need to maintain class-level `partitionReleaseStrategyFactory`.

Reply:
I think this is not possible because the `JobGraph` is attached to the `ExecutionGraph` in an extra step. `ExecutionGraphToSchedulingTopologyAdapter` adapts the `ExecutionGraph` eagerly.

Reply:
Yes, I have not looked through the details in `ExecutionGraphToSchedulingTopologyAdapter` before, only saw it relies `ExecutionGraph`, actually it relies some infos after attaching `JobGraph`.